codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ServerUpgradeCodec.java

```diff
@@ -151,8 +148,23 @@ public boolean prepareUpgradeResponse(ChannelHandlerContext ctx, FullHttpRequest
 
     @Override
     public void upgradeTo(final ChannelHandlerContext ctx, FullHttpRequest upgradeRequest) {
-        // Add the HTTP/2 connection handler to the pipeline immediately following the current handler.
-        ctx.pipeline().addAfter(ctx.name(), handlerName, upgradeToHandler);
+        try {
+            // Add the HTTP/2 connection handler to the pipeline immediately following the current handler.
+            ctx.pipeline().addAfter(ctx.name(), handlerName, connectionHandler);
+            connectionHandler.onHttpServerUpgrade(settings);
+
+        } catch (Http2Exception e) {
+            ctx.fireExceptionCaught(e);
+            ctx.close();
+            return;
+        }
+
+        if (handlers != null) {
+            final String name = ctx.pipeline().context(connectionHandler).name();
```

Comment:
Can handlers be looked up by identity?  Also, what if there are multiple handlers with the same name in the pipeline (i.e. an H2 on top of H2 channel would be present twice in the pipe).

Reply:
I think this should be fine as we lookup by instance... I am missing anything ?

Reply:
Nope, I was thinking of a different API.  I thought the name was derived from the class